### PR TITLE
 [Servicelinker] 2022-05-01 new version

### PR DESCRIPTION
**Pull request by xfz11**
## Description

design: https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/1193

## Checklist

- [x] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [ ] The title of the PR is clear and informative
- [ ] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [ ] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [ ] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
        - {Please put the link here}
    - [ ] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

File: src/ServiceLinker/readme.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+``` yaml
+require:
+# readme.azure.noprofile.md is the common configuration file
+  - $(this-folder)/../readme.azure.noprofile.md
+input-file:
+# You need to specify your swagger files here.
+  # - $(repo)/specification/servicelinker/resource-manager/Microsoft.ServiceLinker/stable/2022-05-01/servicelinker.json
+# If the swagger has not been put in the repo, you may uncomment the following line and refer to it locally
+  - $(this-folder)/../APISpecs/servicelinker/resource-manager/Microsoft.ServiceLinker/stable/2022-05-01/servicelinker.json
```
Comment (BethanyZhou) [inline src/ServiceLinker/readme.md:40]: This path should be public available

File: src/ServiceLinker/custom/Get-AzServiceLinkerConfigurationForContainerApp.ps1 [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+# limitations under the License.
+# Code generated by Microsoft (R) AutoRest Code Generator.Changes may cause incorrect behavior and will be lost if the code
+# is regenerated.
+# ----------------------------------------------------------------------------------
+
+<#
+.Synopsis
+list source configurations for a linker.
+.Description
+list source configurations for a linker.
```
Comment (BethanyZhou) [inline src/ServiceLinker/custom/Get-AzServiceLinkerConfigurationForContainerApp.ps1:21]: The description and Synopsis should be more exact to match with cmdlet name.

File: src/ServiceLinker/custom/Get-AzServiceLinkerConfigurationForSpringCloud.ps1 [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    [Microsoft.Azure.PowerShell.Cmdlets.ServiceLinker.Category('Path')]
+    [System.String]
+    # The fully qualified Azure Resource manager identifier of the resource to be connected.
+    ${ResourceUri},
+
+    [Parameter(Mandatory)]
+    [Microsoft.Azure.PowerShell.Cmdlets.ServiceLinker.Category('Runtime')]
+    [System.String]
+    # The Name of spring cloud service to be connected.
+    ${Service},
```
Comment (BethanyZhou) [inline src/ServiceLinker/custom/Get-AzServiceLinkerConfigurationForSpringCloud.ps1:52]: Why we do not name Service -> ServiceName App->AppName Deployment->DeploymentName

File: src/ServiceLinker/readme.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    where: 
+      - $.paths["/{resourceUri}/providers/Microsoft.ServiceLinker/linkers/{linkerName}"].put.parameters
+      - $.paths["/{resourceUri}/providers/Microsoft.ServiceLinker/linkers/{linkerName}"].patch.parameters
+    transform: >-
+      $[4] = {
+            "name": "x-ms-serviceconnector-user-token",
+            "type": "string",
+            "in": "header"
+          }
+  # - from: swagger-document
```
Comment (BethanyZhou) [inline src/ServiceLinker/readme.md:73]: Please remove unnecessary comments code.

File: src/ServiceLinker/readme.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+### AutoRest Configuration
+> see https://aka.ms/autorest
+
+``` yaml
+require:
+# readme.azure.noprofile.md is the common configuration file
+  - $(this-folder)/../readme.azure.noprofile.md
+input-file:
+# You need to specify your swagger files here.
```
Comment (BethanyZhou) [inline src/ServiceLinker/readme.md:37]: Please add commit id here to lock the version of servicelinker.json. Please refer to: https://github.com/Azure/azure-powershell/blob/12c25f919424b736555b5fd6c2a59e01bef64b9c/src/ContainerInstance/README.md?plain=1#L33

File: src/ServiceLinker/custom/Get-AzServiceLinkerConfigurationForContainerApp.ps1 (hunk 2 above)
Comment (xfz11) [reply src/ServiceLinker/custom/Get-AzServiceLinkerConfigurationForContainerApp.ps1]: updated

File: src/ServiceLinker/readme.md (hunk 4 above)
Comment (xfz11) [reply src/ServiceLinker/readme.md]: removed

File: src/ServiceLinker/readme.md (hunk 5 above)
Comment (xfz11) [reply src/ServiceLinker/readme.md:38]: added

**Review by BethanyZhou — DISMISSED** [review]
LGTM. @wyunchi-ms could you help take care of this PR and move it to main branch?

**Comment by xfz11** [comment]
> LGTM. @wyunchi-ms could you help take care of this PR and move it to main branch?

Thanks Beisi. Could you please hold the PR for a few days? Spring cloud may change module name, we would like to update springcloud related cmdlet name later.

**Comment by BethanyZhou** [comment]
Sure, please let me know this PR is ready to merge by removing label 'DO NOT MERGE'

**Comment by xfz11** [comment]
> Sure, please let me know this PR is ready to merge by removing label 'DO NOT MERGE'

Confirm that spring cloud will not change its cmdlet now. It's ok to merge it. Thanks.

File: src/ServiceLinker/custom/Get-AzServiceLinkerConfigurationForWebApp.ps1 [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+list source configurations for a linker in webapp.
+.Description
+list source configurations for a linker in webapp.
+.Example
+{{ Add code here }}
+.Example
+{{ Add code here }}
+
+.Outputs
+Microsoft.Azure.PowerShell.Cmdlets.ServiceLinker.Models.Api20220501.ISourceConfiguration
```
Comment (wyunchi-ms) [inline src/ServiceLinker/custom/Get-AzServiceLinkerConfigurationForWebApp.ps1:28]: Please remove `Example` and `Outputs` from custom files.

File: src/ServiceLinker/custom/New-AzServiceLinkerForContainerApp.ps1 [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+Create or update linker resource in container app.
+.Description
+Create or update linker resource in container app.
+.Example
+{{ Add code here }}
+.Example
+{{ Add code here }}
+
+.Outputs
+Microsoft.Azure.PowerShell.Cmdlets.ServiceLinker.Models.Api20220501.ILinkerResource
```
Comment (wyunchi-ms) [inline src/ServiceLinker/custom/New-AzServiceLinkerForContainerApp.ps1:28]: Please remove `Example` and `Outputs` from custom files.

File: src/ServiceLinker/custom/New-AzServiceLinkerForSpringCloud.ps1 [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+Create or update linker resource in spring cloud.
+.Description
+Create or update linker resource in spring cloud.
+.Example
+{{ Add code here }}
+.Example
+{{ Add code here }}
+
+.Outputs
+Microsoft.Azure.PowerShell.Cmdlets.ServiceLinker.Models.Api20220501.ILinkerResource
```
Comment (wyunchi-ms) [inline src/ServiceLinker/custom/New-AzServiceLinkerForSpringCloud.ps1:28]: Please remove `Example` and `Outputs` from custom files.

File: src/ServiceLinker/custom/Remove-AzServiceLinkerForContainerApp.ps1 [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+System.Boolean
+.Notes
+COMPLEX PARAMETER PROPERTIES
+
+To create the parameters described below, construct a hash table containing the appropriate properties. For information on hash tables, run Get-Help about_Hash_Tables.
+
+INPUTOBJECT <IServiceLinkerIdentity>: Identity Parameter
+  [Id <String>]: Resource identity path
+  [LinkerName <String>]: The name Linker resource.
+  [ResourceUri <String>]: The fully qualified Azure Resource manager identifier of the resource to be connected.
```
Comment (wyunchi-ms) [inline src/ServiceLinker/custom/Remove-AzServiceLinkerForContainerApp.ps1:39]: Please remove `Example` and `Outputs` from custom files.

File: src/ServiceLinker/custom/Remove-AzServiceLinkerForWebApp.ps1 [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+System.Boolean
+.Notes
+COMPLEX PARAMETER PROPERTIES
+
+To create the parameters described below, construct a hash table containing the appropriate properties. For information on hash tables, run Get-Help about_Hash_Tables.
+
+INPUTOBJECT <IServiceLinkerIdentity>: Identity Parameter
+  [Id <String>]: Resource identity path
+  [LinkerName <String>]: The name Linker resource.
+  [ResourceUri <String>]: The fully qualified Azure Resource manager identifier of the resource to be connected.
```
Comment (wyunchi-ms) [inline src/ServiceLinker/custom/Remove-AzServiceLinkerForWebApp.ps1:39]: Please remove `Example` and `Outputs` from custom files.

File: src/ServiceLinker/custom/Test-AzServiceLinkerForWebApp.ps1 [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+Microsoft.Azure.PowerShell.Cmdlets.ServiceLinker.Models.Api20220501.IValidateResult
+.Notes
+COMPLEX PARAMETER PROPERTIES
+
+To create the parameters described below, construct a hash table containing the appropriate properties. For information on hash tables, run Get-Help about_Hash_Tables.
+
+INPUTOBJECT <IServiceLinkerIdentity>: Identity Parameter
+  [Id <String>]: Resource identity path
+  [LinkerName <String>]: The name Linker resource.
+  [ResourceUri <String>]: The fully qualified Azure Resource manager identifier of the resource to be connected.
```
Comment (wyunchi-ms) [inline src/ServiceLinker/custom/Test-AzServiceLinkerForWebApp.ps1:39]: Please remove `Example` and `Outputs` from custom files.

File: src/ServiceLinker/custom/Update-AzServiceLinkerForSpringCloud.ps1 [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+AUTHINFO <IAuthInfoBase>: The authentication type.
+  AuthType <AuthType>: The authentication type.
+
+INPUTOBJECT <IServiceLinkerIdentity>: Identity Parameter
+  [Id <String>]: Resource identity path
+  [LinkerName <String>]: The name Linker resource.
+  [ResourceUri <String>]: The fully qualified Azure Resource manager identifier of the resource to be connected.
+
+TARGETSERVICE <ITargetServiceBase>: The target service properties
+  Type <Type>: The target service type.
```
Comment (wyunchi-ms) [inline src/ServiceLinker/custom/Update-AzServiceLinkerForSpringCloud.ps1:45]: Please remove `Example` and `Outputs` from custom files.

File: src/ServiceLinker/custom/Update-AzServiceLinkerForWebApp.ps1 [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+AUTHINFO <IAuthInfoBase>: The authentication type.
+  AuthType <AuthType>: The authentication type.
+
+INPUTOBJECT <IServiceLinkerIdentity>: Identity Parameter
+  [Id <String>]: Resource identity path
+  [LinkerName <String>]: The name Linker resource.
+  [ResourceUri <String>]: The fully qualified Azure Resource manager identifier of the resource to be connected.
+
+TARGETSERVICE <ITargetServiceBase>: The target service properties
+  Type <Type>: The target service type.
```
Comment (wyunchi-ms) [inline src/ServiceLinker/custom/Update-AzServiceLinkerForWebApp.ps1:45]: Please remove `Example` and `Outputs` from custom files.

File: src/ServiceLinker/custom/New-AzServiceLinkerUserAssignedIdentityAuthInfoObject.ps1 [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+<#
+.Synopsis
+Create an in-memory object for UserAssignedIdentityAuthInfo.
+.Description
+Create an in-memory object for UserAssignedIdentityAuthInfo.
+
+.Outputs
+Microsoft.Azure.PowerShell.Cmdlets.ServiceLinker.Models.Api20220501.UserAssignedIdentityAuthInfo
+.Link
+https://docs.microsoft.com/powershell/module/az.ServiceLinker/new-AzServiceLinkerUserAssignedIdentityAuthInfoObject
```
Comment (wyunchi-ms) [inline src/ServiceLinker/custom/New-AzServiceLinkerUserAssignedIdentityAuthInfoObject.ps1:26]: Please use lower case in the http link.

File: src/ServiceLinker/custom/New-AzServiceLinkerSystemAssignedIdentityAuthInfoObject.ps1 [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+<#
+.Synopsis
+Create an in-memory object for SystemAssignedIdentityAuthInfo.
+.Description
+Create an in-memory object for SystemAssignedIdentityAuthInfo.
+
+.Outputs
+Microsoft.Azure.PowerShell.Cmdlets.ServiceLinker.Models.Api20220501.SystemAssignedIdentityAuthInfo
+.Link
+https://docs.microsoft.com/powershell/module/az.ServiceLinker/new-AzServiceLinkerSystemAssignedIdentityAuthInfoObject
```
Comment (wyunchi-ms) [inline src/ServiceLinker/custom/New-AzServiceLinkerSystemAssignedIdentityAuthInfoObject.ps1:26]: Please use lower case in the http link.

File: src/ServiceLinker/custom/New-AzServiceLinkerKeyVaultSecretReferenceSecretInfoObject.ps1 [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+<#
+.Synopsis
+Create an in-memory object for KeyVaultSecretReferenceSecretInfo.
+.Description
+Create an in-memory object for KeyVaultSecretReferenceSecretInfo.
+
+.Outputs
+Microsoft.Azure.PowerShell.Cmdlets.ServiceLinker.Models.Api20220501.KeyVaultSecretReferenceSecretInfo
+.Link
+https://docs.microsoft.com/powershell/module/az.ServiceLinker/new-AzServiceLinkerKeyVaultSecretReferenceSecretInfoObject
```
Comment (wyunchi-ms) [inline src/ServiceLinker/custom/New-AzServiceLinkerKeyVaultSecretReferenceSecretInfoObject.ps1:26]: Please use lower case in the http link.

File: src/ServiceLinker/custom/New-AzServiceLinkerKeyVaultSecretUriSecretInfoObject.ps1 [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+<#
+.Synopsis
+Create an in-memory object for KeyVaultSecretUriSecretInfo.
+.Description
+Create an in-memory object for KeyVaultSecretUriSecretInfo.
+
+.Outputs
+Microsoft.Azure.PowerShell.Cmdlets.ServiceLinker.Models.Api20220501.KeyVaultSecretUriSecretInfo
+.Link
+https://docs.microsoft.com/powershell/module/az.ServiceLinker/new-AzServiceLinkerKeyVaultSecretUriSecretInfoObject
```
Comment (wyunchi-ms) [inline src/ServiceLinker/custom/New-AzServiceLinkerKeyVaultSecretUriSecretInfoObject.ps1:26]: Please use lower case in the http link.

File: src/ServiceLinker/custom/New-AzServiceLinkerSecretAuthInfoObject.ps1 [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+<#
+.Synopsis
+Create an in-memory object for SecretAuthInfo.
+.Description
+Create an in-memory object for SecretAuthInfo.
+
+.Outputs
+Microsoft.Azure.PowerShell.Cmdlets.ServiceLinker.Models.Api20220501.SecretAuthInfo
+.Link
+https://docs.microsoft.com/powershell/module/az.ServiceLinker/new-AzServiceLinkerSecretAuthInfoObject
```
Comment (wyunchi-ms) [inline src/ServiceLinker/custom/New-AzServiceLinkerSecretAuthInfoObject.ps1:26]: Please use lower case in the http link.

File: src/ServiceLinker/custom/New-AzServiceLinkerServicePrincipalSecretAuthInfoObject.ps1 [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+<#
+.Synopsis
+Create an in-memory object for ServicePrincipalSecretAuthInfo.
+.Description
+Create an in-memory object for ServicePrincipalSecretAuthInfo.
+
+.Outputs
+Microsoft.Azure.PowerShell.Cmdlets.ServiceLinker.Models.Api20220501.ServicePrincipalSecretAuthInfo
+.Link
+https://docs.microsoft.com/powershell/module/az.ServiceLinker/new-AzServiceLinkerServicePrincipalSecretAuthInfoObject
```
Comment (wyunchi-ms) [inline src/ServiceLinker/custom/New-AzServiceLinkerServicePrincipalSecretAuthInfoObject.ps1:26]: Please use lower case in the http link.

File: src/ServiceLinker/custom/New-AzServiceLinkerValueSecretInfoObject.ps1 [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+<#
+.Synopsis
+Create an in-memory object for ValueSecretInfo.
+.Description
+Create an in-memory object for ValueSecretInfo.
+
+.Outputs
+Microsoft.Azure.PowerShell.Cmdlets.ServiceLinker.Models.Api20220501.ValueSecretInfo
+.Link
+https://docs.microsoft.com/powershell/module/az.ServiceLinker/new-AzServiceLinkerValueSecretInfoObject
```
Comment (wyunchi-ms) [inline src/ServiceLinker/custom/New-AzServiceLinkerValueSecretInfoObject.ps1:26]: Please use lower case in the http link.

File: src/ServiceLinker/custom/Get-AzServiceLinkerConfigurationForContainerApp.ps1 [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+list source configurations for a linker in container app.
+.Description
+list source configurations for a linker in container app.
+.Example
+{{ Add code here }}
+.Example
+{{ Add code here }}
+
+.Outputs
+Microsoft.Azure.PowerShell.Cmdlets.ServiceLinker.Models.Api20220501.ISourceConfiguration
```
[the remark this answers is not in the document]
Comment (xfz11) [reply src/ServiceLinker/custom/Get-AzServiceLinkerConfigurationForContainerApp.ps1]: removed

File: src/ServiceLinker/custom/Get-AzServiceLinkerConfigurationForWebApp.ps1 (hunk 6 above)
Comment (xfz11) [reply src/ServiceLinker/custom/Get-AzServiceLinkerConfigurationForWebApp.ps1]: removed

File: src/ServiceLinker/custom/Get-AzServiceLinkerForContainerApp.ps1 [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+Microsoft.Azure.PowerShell.Cmdlets.ServiceLinker.Models.Api20220501.ILinkerResource
+.Notes
+COMPLEX PARAMETER PROPERTIES
+
+To create the parameters described below, construct a hash table containing the appropriate properties. For information on hash tables, run Get-Help about_Hash_Tables.
+
+INPUTOBJECT <IServiceLinkerIdentity>: Identity Parameter
+  [Id <String>]: Resource identity path
+  [LinkerName <String>]: The name Linker resource.
+  [ResourceUri <String>]: The fully qualified Azure Resource manager identifier of the resource to be connected.
```
[the remark this answers is not in the document]
Comment (xfz11) [reply src/ServiceLinker/custom/Get-AzServiceLinkerForContainerApp.ps1]: removed

File: src/ServiceLinker/custom/New-AzServiceLinkerKeyVaultSecretReferenceSecretInfoObject.ps1 (hunk 16 above)
Comment (xfz11) [reply src/ServiceLinker/custom/New-AzServiceLinkerKeyVaultSecretReferenceSecretInfoObject.ps1]: updated

**Review by wyunchi-ms — APPROVED** [review]

**Pull request merged**